### PR TITLE
compress: Fix unsound optimization of `new RegExp(…)`

### LINKED ISSUE #6941: Minifier should not drop `new` for `RegExp`
### Describe the bug

I'm not sure if this is important, though.

Reported by https://github.com/vercel/next.js/discussions/30237#discussioncomment-4961296

### Input code

```typescript
const pattern = /[abc]+/g;

const instance1 = RegExp(pattern);
const instance2 = new RegExp(pattern);

instance1 === pattern // true
instance 2 === pattern // false
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": true
    },
    "target": "es2020",
    "loose": false,
    "minify": {
      "compress": false,
      "mangle": false
    }
  },
  "module": {
    "type": "es6"
  },
  "minify": false,
  "isModule": true
}
```

### Playground link

https://play.swc.rs/?version=1.3.35&code=H4sIAAAAAAAAA0vOzysuUShILClJLcpTsFXQj05MSo7V1k%2B35uJKBstlAonEvORUQ6BsUGq6a0WBBlS5pjWaEiOgkrzUckxlXEiG2NrCrdPXVygpKk2FyyoYoUunJeYUpwIAjgX8hqQAAAA%3D&config=H4sIAAAAAAAAA0WOSwrDMAxE76J1FiGLLnyHHsK4SnDxD0mBGuO7VzYJ2Q0zo6dp8GUHpkGxxEhDcU1if2BAakF25IvAAsJqCZ3YVVs6ULSBvK3bqmnImRHMbgPjAtEnv9fBcjkWQuYnsukId7MrK%2BbPOYw2303mC%2FrDuO48v6%2Fi3PAHuM3Ro7cAAAA%3D

### Expected behavior

It should pass

### Actual behavior

_No response_

### Version

1.3.35

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by andersk**
In #4932, an optimization was added that transforms `new RegExp(…)` to `RegExp(…)`.  It seems to have been motivated by this sentence in the ECMAScript specification:

> Thus the function call `RegExp(…)` is equivalent to the object creation expression `new RegExp(…)` with the same arguments.

But that sentence turns out to be wrong.  It contradicts the detailed steps in the following section of the specification, as well as the empirical behavior of JavaScriptCore, SpiderMonkey, and V8:

    > r = /a/;
    /a/
    > RegExp(r) === r
    true
    > new RegExp(r) === r
    false

- tc39/ecma262#3028

Restrict this optimization to the case where we can prove one of the first two arguments is a string.

Fixes #6941.

### CODE REVIEW

File: crates/swc_ecma_minifier/src/compress/pure/misc.rs [hunk 1]
```diff
@@ -16,6 +16,46 @@ use crate::compress::{
...
+        Expr::Paren(ParenExpr { expr, .. }) => is_definitely_string(expr),
+        _ => false,
+    }
+}
+
+/// Check whether we can compress `new RegExp(…)` to `RegExp(…)`. That's sound
+/// unless the first argument is already a RegExp object and the second is
+/// undefined. We check for the common case where we can prove one of the first
+/// two arguments is a string.
+fn can_compress_new_regexp(args: &Option<Vec<ExprOrSpread>>) -> bool {
```
Comment (kdy1) [inline crates/swc_ecma_minifier/src/compress/pure/misc.rs:39]:
```suggestion
fn can_compress_new_regexp(args: Option<&[ExprOrSpread]>) -> bool {
```

File: crates/swc_ecma_minifier/src/compress/pure/misc.rs [hunk 2]
```diff
@@ -569,7 +607,8 @@ impl Pure<'_> {
                     "TypeError",
                     "URIError",
                 ],
-            ) =>
+            ) || (callee.is_global_ref_to(&self.expr_ctx, "RegExp")
+                && can_compress_new_regexp(args)) =>
```
Comment (kdy1) [inline crates/swc_ecma_minifier/src/compress/pure/misc.rs:611]:
```suggestion
                && can_compress_new_regexp(args.as_deref())) =>
```

**Review by kdy1 — CHANGES_REQUESTED** [review]
swc-bump: 
 - swc_ecma_minifier

File: crates/swc_ecma_minifier/src/compress/pure/misc.rs (hunk 1 above)
Comment (andersk) [reply crates/swc_ecma_minifier/src/compress/pure/misc.rs]: Done.

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_minifier

**Pull request merged**